File: src/Api/Services/SeriesService.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using PostGenerator.Api.Data;
using PostGenerator.Api.Models;
using PostGenerator.Core;

namespace PostGenerator.Api.Services;

public class SeriesService : ISeriesService
{
    private readonly AppDbContext _db;
    private readonly IPostGenerationService _postGeneration;

    public SeriesService(AppDbContext db, IPostGenerationService postGeneration)
    {
        _db = db;
        _postGeneration = postGeneration;
    }

    public async Task<(int SeriesId, IReadOnlyList<int> PostIds)?> GenerateAsync(int userId, GenerateSeriesRequest request, CancellationToken cancellationToken = default)
    {
        if (!Enum.TryParse<PostPlatform>(request.Platform, true, out var platform))
            return null;

        var options = new GeneratePostOptions(request.TopicDetail, request.NumPosts, request.Platform, request.Linked, request.Tone, request.Length, request.TikTokScriptDurationSeconds);
        var generated = await _postGeneration.GeneratePostsAsync(options, cancellationToken);
        if (generated.Count == 0) return null;

        var series = new PostSeries
        {
            UserId = userId,
            TopicDetail = request.TopicDetail,
            NumPosts = generated.Count,
            OptionsJson = JsonSerializer.Serialize(new { request.Linked, request.Tone, request.Length, request.GenerateImages, request.StartDate, request.Recurrence, request.ScheduledTimeOfDay }),
            CreatedAt = DateTime.UtcNow,
        };
        _db.PostSeriesSet.Add(series);
        await _db.SaveChangesAsync(cancellationToken);

        var startDate = request.StartDate ?? DateTime.UtcNow.Date;
        var timeOfDay = string.IsNullOrEmpty(request.ScheduledTimeOfDay) ? (TimeSpan?)null : (TimeOnly.TryParse(request.ScheduledTimeOfDay, out var t) ? t.ToTimeSpan() : null);
        var posts = new List<Post>();
        for (var i = 0; i < generated.Count; i++)
        {
            var g = generated[i];
            DateTime? scheduledAt = null;
            if (request.StartDate.HasValue && timeOfDay.HasValue)
                scheduledAt = startDate.Add(timeOfDay.Value).AddDays(i * (request.Recurrence == "daily" ? 1 : 7));
            posts.Add(new Post
            {
                UserId = userId,
                TopicSummary = request.TopicDetail.Length > 500 ? request.TopicDetail[..500] : request.TopicDetail,
                Platform = platform,
                Status = scheduledAt.HasValue ? PostStatus.Scheduled : PostStatus.Draft,
                ScheduledAt = scheduledAt,
                Content = g.Content,
                Script = g.Script,
                MetadataJson = g.HashtagsJson != null ? JsonSerializer.Serialize(new { hashtags = JsonSerializer.Deserialize<object>(g.HashtagsJson) }) : null,
                Tone = request.Tone,
                Length = request.Length,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow,
            });
        }
        _db.Posts.AddRange(posts);
        await _db.SaveChangesAsync(cancellationToken);
        return (series.Id, posts.Select(p => p.Id).ToList());
    }

    public async Task GenerateStreamAsync(int userId, GenerateSeriesRequest request, Func<int, PostDto, CancellationToken, Task> onPost, CancellationToken cancellationToken = default)
    {
        if (!Enum.TryParse<PostPlatform>(request.Platform, true, out var platform))
            throw new ArgumentException("Invalid platform", nameof(request));

        var series = new PostSeries
        {
            UserId = userId,
            TopicDetail = request.TopicDetail,
            NumPosts = request.NumPosts,
            OptionsJson = JsonSerializer.Serialize(new { request.Linked, request.Tone, request.Length, request.GenerateImages, request.StartDate, request.Recurrence, request.ScheduledTimeOfDay }),
            CreatedAt = DateTime.UtcNow,
        };
        _db.PostSeriesSet.Add(series);
        await _db.SaveChangesAsync(cancellationToken);
        var seriesId = series.Id;

        var options = new GeneratePostOptions(request.TopicDetail, request.NumPosts, request.Platform, request.Linked, request.Tone, request.Length, request.TikTokScriptDurationSeconds);

        var startDate = request.StartDate ?? DateTime.UtcNow.Date;
        var timeOfDay = string.IsNullOrEmpty(request.ScheduledTimeOfDay) ? (TimeSpan?)null : (TimeOnly.TryParse(request.ScheduledTimeOfDay, out var t) ? t.ToTimeSpan() : null);
        var topicSummary = request.TopicDetail.Length > 500 ? request.TopicDetail[..500] : request.TopicDetail;
        var previousContents = new List<string>();

        for (var i = 0; i < request.NumPosts; i++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var generated = await _postGeneration.GenerateSinglePostAsync(options, i + 1, previousContents, cancellationToken);
            if (generated == null) throw new InvalidOperationException("Post generation returned null.");

            DateTime? scheduledAt = null;
            if (request.StartDate.HasValue && timeOfDay.HasValue)
                scheduledAt = startDate.Add(timeOfDay.Value).AddDays(i * (request.Recurrence == "daily" ? 1 : 7));

            var status = scheduledAt.HasValue ? PostStatus.Scheduled : PostStatus.Draft;
            var now = DateTime.UtcNow;
            var metadataJson = generated.HashtagsJson != null ? JsonSerializer.Serialize(new { hashtags = JsonSerializer.Deserialize<object>(generated.HashtagsJson) }) : null;

            var dto = new PostDto(
                Id: 0,
                UserId: userId,
                TopicSummary: topicSummary,
                Platform: platform.ToString(),
                Status: status.ToString(),
                ScheduledAt: scheduledAt,
                PublishedAt: null,
                ExternalPostId: null,
                ViewsCount: null,
                LikesCount: null,
                CommentsCount: null,
                LastEngagementFetchedAt: null,
                Content: generated.Content,
                Script: generated.Script,
                ImageUrl: null,
                MetadataJson: metadataJson,
                Tone: request.Tone,
                Length: request.Length,
                CreatedAt: now,
                UpdatedAt: now);

            await onPost(seriesId, dto, cancellationToken);

            previousContents.Add(generated.Content);
        }
    }

    public async Task<(int SeriesId, IReadOnlyList<int> PostIds)> PublishGeneratedSeriesAsync(int userId, PublishGeneratedSeriesRequest request, CancellationToken cancellationToken = default)
    {
        if (!Enum.TryParse<PostPlatform>(request.Platform, true, out var platform))
            throw new ArgumentException("Invalid platform", nameof(request));

        var series = new PostSeries
        {
            UserId = userId,
            TopicDetail = request.TopicDetail,
            NumPosts = request.GeneratedPosts.Count,
            OptionsJson = JsonSerializer.Serialize(new { request.Linked, request.Tone, request.Length, request.GenerateImages, request.StartDate, request.Recurrence, request.ScheduledTimeOfDay }),
            CreatedAt = DateTime.UtcNow,
        };
        _db.PostSeriesSet.Add(series);
        await _db.SaveChangesAsync(cancellationToken);

        var startDate = request.StartDate ?? DateTime.UtcNow.Date;
        var timeOfDay = string.IsNullOrEmpty(request.ScheduledTimeOfDay) ? (TimeSpan?)null : (TimeOnly.TryParse(request.ScheduledTimeOfDay, out var t) ? t.ToTimeSpan() : null);
        var topicSummary = request.TopicDetail.Length > 500 ? request.TopicDetail[..500] : request.TopicDetail;
        var posts = new List<Post>();

        for (var i = 0; i < request.GeneratedPosts.Count; i++)
        {
            var item = request.GeneratedPosts[i];
            DateTime? scheduledAt = null;
            if (request.StartDate.HasValue && timeOfDay.HasValue)
                scheduledAt = startDate.Add(timeOfDay.Value).AddDays(i * (request.Recurrence == "daily" ? 1 : 7));

            posts.Add(new Post
            {
                UserId = userId,
                TopicSummary = topicSummary,
                Platform = platform,
                Status = scheduledAt.HasValue ? PostStatus.Scheduled : PostStatus.Draft,
                ScheduledAt = scheduledAt,
                Content = item.Content,
                Script = item.Script,
                ImageUrl = item.ImageUrl,
                MetadataJson = item.MetadataJson,
                Tone = request.Tone,
                Length = request.Length,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow,
            });
        }

        _db.Posts.AddRange(posts);
        await _db.SaveChangesAsync(cancellationToken);
        return (series.Id, posts.Select(p => p.Id).ToList());
    }
}

File: src/lib/services/series-service.ts
import { db } from "@/db/drizzle";
import { posts, postSeries } from "@/db/schema";
import type { PostPlatform } from "@/db/schema";
import {
  generatePosts,
  generateSinglePost,
  type GeneratePostOptions,
  type GeneratedPostItem,
} from "./post-generation-service";
import type { PostDto } from "./post-service";

const PLATFORMS = [
  "LinkedIn",
  "Skool",
  "Instagram",
  "Bluesky",
  "Facebook",
  "TikTok",
] as const;

function isValidPlatform(s: string): s is PostPlatform {
  return PLATFORMS.includes(s as PostPlatform);
}

export interface GenerateSeriesRequest {
  topicDetail: string;
  numPosts: number;
  platform: string;
  linked?: boolean;
  tone?: string | null;
  length?: string | null;
  generateImages?: boolean;
  tiktokScriptDurationSeconds?: number | null;
  startDate?: Date | null;
  recurrence?: string | null;
  scheduledTimeOfDay?: string | null;
}

export interface PublishGeneratedRequest {
  topicDetail: string;
  numPosts: number;
  platform: string;
  generatedPosts: {
    content: string;
    script?: string | null;
    metadataJson?: string | null;
    imageUrl?: string | null;
  }[];
  linked?: boolean;
  tone?: string | null;
  length?: string | null;
  generateImages?: boolean;
  tiktokScriptDurationSeconds?: number | null;
  startDate?: Date | null;
  recurrence?: string | null;
  scheduledTimeOfDay?: string | null;
}

function parseTimeOfDay(s: string | null | undefined): number | null {
  if (!s) return null;
  const match = s.match(/^(\d{1,2}):(\d{2})$/);
  if (!match) return null;
  const h = parseInt(match[1], 10);
  const m = parseInt(match[2], 10);
  if (h < 0 || h > 23 || m < 0 || m > 59) return null;
  return h * 60 + m;
}

export async function generate(
  userId: number,
  req: GenerateSeriesRequest
): Promise<{ seriesId: number; postIds: number[] } | null> {
  if (!isValidPlatform(req.platform)) return null;

  const options: GeneratePostOptions = {
    topicDetail: req.topicDetail,
    numPosts: req.numPosts,
    platform: req.platform,
    linked: req.linked ?? false,
    tone: req.tone,
    length: req.length,
    tiktokScriptDurationSeconds: req.tiktokScriptDurationSeconds ?? null,
  };

  const generated = await generatePosts(options);
  if (generated.length === 0) return null;

  const [series] = await db
    .insert(postSeries)
    .values({
      userId,
      topicDetail: req.topicDetail,
      numPosts: generated.length,
      optionsJson: JSON.stringify({
        linked: req.linked,
        tone: req.tone,
        length: req.length,
        generateImages: req.generateImages,
        startDate: req.startDate,
        recurrence: req.recurrence,
        scheduledTimeOfDay: req.scheduledTimeOfDay,
      }),
    })
    .returning();

  if (!series) return null;

  const startDate = req.startDate ? new Date(req.startDate) : new Date();
  const timeMins = parseTimeOfDay(req.scheduledTimeOfDay);
  const recurrence = req.recurrence === "daily" ? 1 : 7;

  const postRows = [];
  for (let i = 0; i < generated.length; i++) {
    const g = generated[i];
    let scheduledAt: Date | null = null;
    if (req.startDate && timeMins != null) {
      const d = new Date(startDate);
      d.setHours(Math.floor(timeMins / 60), timeMins % 60, 0, 0);
      d.setDate(d.getDate() + i * recurrence);
      scheduledAt = d;
    }

    postRows.push({
      userId,
      topicSummary:
        req.topicDetail.length > 500
          ? req.topicDetail.slice(0, 500)
          : req.topicDetail,
      platform: req.platform as PostPlatform,
      status: scheduledAt ? ("Scheduled" as const) : ("Draft" as const),
      scheduledAt,
      content: g.content,
      script: g.script,
      metadataJson: g.hashtagsJson
        ? JSON.stringify({ hashtags: JSON.parse(g.hashtagsJson) })
        : null,
      tone: req.tone,
      length: req.length,
    });
  }

  const inserted = await db.insert(posts).values(postRows).returning();
  return {
    seriesId: series.id,
    postIds: inserted.map((p) => p.id),
  };
}

export async function generateStream(
  userId: number,
  req: GenerateSeriesRequest,
  onPost: (seriesId: number, post: PostDto) => Promise<void>
): Promise<void> {
  if (!isValidPlatform(req.platform)) {
    throw new Error("Invalid platform");
  }

  const [series] = await db
    .insert(postSeries)
    .values({
      userId,
      topicDetail: req.topicDetail,
      numPosts: req.numPosts,
      optionsJson: JSON.stringify({
        linked: req.linked,
        tone: req.tone,
        length: req.length,
        generateImages: req.generateImages,
        startDate: req.startDate,
        recurrence: req.recurrence,
        scheduledTimeOfDay: req.scheduledTimeOfDay,
      }),
    })
    .returning();

  if (!series) throw new Error("Failed to create series");

  const options: GeneratePostOptions = {
    topicDetail: req.topicDetail,
    numPosts: req.numPosts,
    platform: req.platform,
    linked: req.linked ?? false,
    tone: req.tone,
    length: req.length,
    tiktokScriptDurationSeconds: req.tiktokScriptDurationSeconds ?? null,
  };

  const startDate = req.startDate ? new Date(req.startDate) : new Date();
  const timeMins = parseTimeOfDay(req.scheduledTimeOfDay);
  const recurrence = req.recurrence === "daily" ? 1 : 7;
  const topicSummary =
    req.topicDetail.length > 500 ? req.topicDetail.slice(0, 500) : req.topicDetail;
  const previousContents: string[] = [];

  for (let i = 0; i < req.numPosts; i++) {
    const generated = await generateSinglePost(options, i + 1, previousContents);

    let scheduledAt: Date | null = null;
    if (req.startDate && timeMins != null) {
      const d = new Date(startDate);
      d.setHours(Math.floor(timeMins / 60), timeMins % 60, 0, 0);
      d.setDate(d.getDate() + i * recurrence);
      scheduledAt = d;
    }

    const status = scheduledAt ? "Scheduled" : "Draft";
    const now = new Date();
    const metadataJson = generated.hashtagsJson
      ? JSON.stringify({ hashtags: JSON.parse(generated.hashtagsJson) })
      : null;

    const postDto: PostDto = {
      id: 0,
      userId,
      topicSummary,
      platform: req.platform,
      status,
      scheduledAt: scheduledAt?.toISOString() ?? null,
      publishedAt: null,
      externalPostId: null,
      viewsCount: null,
      likesCount: null,
      commentsCount: null,
      lastEngagementFetchedAt: null,
      content: generated.content,
      script: generated.script ?? null,
      imageUrl: null,
      metadataJson,
      tone: req.tone ?? null,
      length: req.length ?? null,
      createdAt: now.toISOString(),
      updatedAt: now.toISOString(),
    };

    await onPost(series.id, postDto);
    previousContents.push(generated.content);
  }
}

export async function publishGenerated(
  userId: number,
  req: PublishGeneratedRequest
): Promise<{ seriesId: number; postIds: number[] }> {
  if (!isValidPlatform(req.platform)) {
    throw new Error("Invalid platform");
  }

  const [series] = await db
    .insert(postSeries)
    .values({
      userId,
      topicDetail: req.topicDetail,
      numPosts: req.generatedPosts.length,
      optionsJson: JSON.stringify({
        linked: req.linked,
        tone: req.tone,
        length: req.length,
        generateImages: req.generateImages,
        startDate: req.startDate,
        recurrence: req.recurrence,
        scheduledTimeOfDay: req.scheduledTimeOfDay,
      }),
    })
    .returning();

  if (!series) throw new Error("Failed to create series");

  const startDate = req.startDate ? new Date(req.startDate) : new Date();
  const timeMins = parseTimeOfDay(req.scheduledTimeOfDay);
  const recurrence = req.recurrence === "daily" ? 1 : 7;
  const topicSummary =
    req.topicDetail.length > 500 ? req.topicDetail.slice(0, 500) : req.topicDetail;

  const postRows = req.generatedPosts.map((item, i) => {
    let scheduledAt: Date | null = null;
    if (req.startDate && timeMins != null) {
      const d = new Date(startDate);
      d.setHours(Math.floor(timeMins / 60), timeMins % 60, 0, 0);
      d.setDate(d.getDate() + i * recurrence);
      scheduledAt = d;
    }
    return {
      userId,
      topicSummary,
      platform: req.platform as PostPlatform,
      status: scheduledAt ? ("Scheduled" as const) : ("Draft" as const),
      scheduledAt,
      content: item.content,
      script: item.script ?? null,
      imageUrl: item.imageUrl ?? null,
      metadataJson: item.metadataJson ?? null,
      tone: req.tone ?? null,
      length: req.length ?? null,
    };
  });

  const inserted = await db.insert(posts).values(postRows).returning();
  return {
    seriesId: series.id,
    postIds: inserted.map((p) => p.id),
  };
}
